src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/ActiveWorkspaceProjectContextTracker.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.LanguageServices;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.LanguageServices
+{
+    /// <summary>
+    ///     Provides an implementation of <see cref="IActiveWorkspaceProjectContextTracker"/> that tracks the 
+    ///     "active" context by handling the VSHPROPID_ActiveIntellisenseProjectContext IVsHierarchy property.
+    /// </summary>
+    [Export(typeof(IActiveIntellisenseProjectProvider))]
+    [ExportProjectNodeComService(typeof(IVsContainedLanguageProjectNameProvider))]
+    [AppliesTo(ProjectCapability.DotNetLanguageService2)]
+    internal class ActiveWorkspaceProjectContextTracker : IActiveIntellisenseProjectProvider, IVsContainedLanguageProjectNameProvider, IActiveWorkspaceProjectContextTracker
+    {
+        private ImmutableDictionary<IWorkspaceProjectContext, string> _contexts = ImmutableDictionary<IWorkspaceProjectContext, string>.Empty;
+
+        [ImportingConstructor]
+        public ActiveWorkspaceProjectContextTracker(UnconfiguredProject project) // For scoping
+        {
+        }
+
+        public string ActiveIntellisenseProjectContext
+        {
+            get;
+            set;
```

Comment:
What are the threading rules about setting this value?

Reply:
It's always set on UI thread, but if it's not what concerns do you have? Being out of sync after we've queried it?